config/custom_types.go

```diff
@@ -0,0 +1,189 @@
+	var num uint64
+	if err := value.Decode(&num); err == nil {
+		*b = ByteSize(num)
+		return nil
+	}
+	var s string
+	if err := value.Decode(&s); err == nil {
+		bs, parseErr := parseByteSizeFromString(s)
+		if parseErr != nil {
+			return parseErr
+		}
+		*b = bs
+		return nil
+	}
+	return fmt.Errorf("invalid bytes format: %v", value)
+}
+
+// UnmarshalText allows decoding from text.
+// Implements encoding.TextUnmarshaler interface, which is used by mapstructure.TextUnmarshallerHookFunc.
+func (b *ByteSize) UnmarshalText(text []byte) error {
+	return b.UnmarshalJSON(text)
+}
+
+// String returns the human-readable string representation.
+// Implements fmt.Stringer interface.
+func (b ByteSize) String() string {
+	return bytefmt.ByteSize(uint64(b))
+}
+
+// MarshalJSON encodes as a human-readable string in JSON.
+// Implements json.Marshaler interface.
+func (b ByteSize) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.String())
+}
+
+// MarshalYAML encodes as a human-readable string in YAML.
+// Implements yaml.Marshaler interface.
+func (b ByteSize) MarshalYAML() (interface{}, error) {
+	return b.String(), nil
+}
+
+// MarshalText encodes as a human-readable string in text.
+// Implements encoding.TextMarshaler interface.
+func (b *ByteSize) MarshalText() ([]byte, error) {
+	return []byte(b.String()), nil
+}
+
+func parseByteSizeFromString(s string) (ByteSize, error) {
+	v := strings.TrimSpace(s)
+
+	// Handle k8s power-of-two values.
+	for _, k8sByteSuffix := range [...]string{"Ki", "Mi", "Gi", "Ti", "Pi", "Ei"} {
+		if strings.HasSuffix(v, k8sByteSuffix) {
+			v = v[:len(v)-1]
+			break
+		}
+	}
+
+	num, err := bytefmt.ToBytes(v)
+	if err != nil {
+		return 0, fmt.Errorf("invalid byte size format (%s): %w", s, err)
+	}
+	return ByteSize(num), nil
+}
+
+// TimeDuration represents a time duration that can be parsed from JSON and YAML.
+// This type is intended to be used in configuration structures
+// and allows parsing both integers (nanoseconds) and human-readable strings (e.g. "1h30m").
+type TimeDuration time.Duration
+
+// UnmarshalJSON allows decoding from JSON and supports both integers (nanoseconds) and human-readable strings.
+// Implements json.Unmarshaler interface.
+func (d *TimeDuration) UnmarshalJSON(data []byte) error {
+	s := strings.Trim(string(data), `"`)
+	if num, err := strconv.ParseInt(s, 10, 64); err == nil {
+		if num < 0 {
+			return fmt.Errorf("negative value is not allowed: %d", num)
+		}
+		*d = TimeDuration(num)
+		return nil
+	}
+
+	if dur, err := time.ParseDuration(s); err == nil {
+		*d = TimeDuration(dur)
+		return nil
+	}
+
+	return fmt.Errorf("invalid duration format: %s", s)
+}
+
+// UnmarshalYAML allows decoding from YAML and supports both integers (nanoseconds) and human-readable strings.
+// Implements yaml.Unmarshaler interface.
+func (d *TimeDuration) UnmarshalYAML(value *yaml.Node) error {
+	var raw string
+	if err := value.Decode(&raw); err == nil {
```

Comment:
If the duration format is incorrect, the original error will be lost, and the value will be parsed as an integer instead.
Example: 1N30m
As a result, the error message will appear as:
"Invalid format duration: <val>"
By returning ParseDuration, the error message will be more user-friendly:
"unknown unit 'N' in duration '1N30m'

Reply:
Great point, thanks. Fixed